tests/test_iotedgedev.py

```diff
@@ -203,3 +235,12 @@ def test_load_no_dotenv():
     # assert '.env.test file not found on disk.' in result.output
     # assert 'PROCESSING' in result.output
 '''
+
+
+def add_module_and_verify(main, runner, template):
```

Comment:
I'm currently getting this error during the py27 portion of a tox run. py36 seems to get through it. It might be something wrong on my machine, but please double check.

```
================================== FAILURES ===================================
_______________________________ test_module_add _______________________________

    def test_module_add():
        """Test the addmodule command"""
        os.chdir(test_solution_dir)
        cli = __import__("iotedgedev.cli", fromlist=['main'])
        runner = CliRunner()

>       add_module_and_verify(cli.main, runner, "csharp")

..\test_iotedgedev.py:106:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

main = <click.core.Group object at 0x0000000005C04FD0>
runner = <click.testing.CliRunner object at 0x0000000005C619E8>
template = 'csharp'

    def add_module_and_verify(main, runner, template):
        module_name = template + "module"
        result = runner.invoke(main, ['addmodule', module_name, '--template', template])
        print(result.output)
>       assert 'CREATE COMPLETE' in result.output
E       assert 'CREATE COMPLETE' in '\n=========================================\n======== PROCESSING CONFIG FILES ========\n================================...ed in 858.83 ms for E:\iotedgedev\tests\test_solution\modules\csharpmodule\cshar\nRestore succeeded.\n'
E        +  where '\n=========================================\n======== PROCESSING CONFIG FILES ========\n=============\nRestore succeeded.\n' = <Result IOError(22, "invalid mode ('r') or filename")>.outputarpmodule\csharpmodule.csproj.

..\test_iotedgedev.py:244: AssertionError
```

Reply:
Can you help check if there is a valid DEPLOYMENT_CONFIG_TEMPLATE_FILE key in your .env file? I also added more detailed error messages in this update. Please help run the test again and let me know if the error still exists.

Reply:
The changes to handle this at runtime for the user look very good, and the test is OK if I put DEPLOYMENT_CONFIG_TEMPLATE_FILE in .env.

There is still a test failure when DEPLOYMENT_CONFIG_TEMPLATE_FILE is not present. It seems like `add_module_and_verify` goes directly to `os.environ`, so it doesn't get the default from `envvars`? 

Other than that case, looks good to me.

Reply:
Thanks for the test. I can reproduce it on my side. I fixed it by setting `os.environ` when `get_envvar` falls back to default value. Please try again.

Reply:
This fixes the test, thanks!

I'm not sure that we should mutate `os.environ` in every case that we take a default value, though. I don't have a very strong opinion here. I believe that in python, setting `os.environ` is scoped to the process and child processes, so setting it should not have undesired side effects for the user. 

With that in mind, two alternatives to consider.

One would be to have `test_iotedgedev` import and load `envvars`. Then the code in `add_module_and_verify` can look like 
`    assert module_name in json.load(open(envvars.DEPLOYMENT_CONFIG_TEMPLATE_FILE))["moduleContent"]["$edgeAgent"]["properties.desired"]["modules"]`, 

With this approach, no change to the `envvars` behavior is required. The test would benefit from the existing `load` setting the default value on `self`, and we avoid touching the underlying `os.environ`. 

Since `test_iotedgedev` runs end-to-end tests, I'm comfortable with the direct dependency on `envvars`. However, if we wanted to avoid that direct dependency, a second alternative might be to promote our default value to a constant. Then `test_iotedgedev` can check `os.environ`, but fallback to the constant.

Reply:
Yes...Actually I considered alternative solutions including the ones you proposed but failed to find a strong reason to prefer one over another so I simply chose the easiest one :) Thanks for your kind suggestions anyway.